### PR TITLE
feat: harvester

### PR DESCRIPTION
**Pull request by WuJun2016**
merge the Harvester UI into the rancher dashboard
- for multi-cluster view, harvester UI is accessed via the `virtualization management` menu by clicking one of the harvester clusters.
- for a single harvester cluster view, it will default to go the local Harvester cluster, and all other rancher-related pages are hidden.

### CODE REVIEW

File: assets/brand/harvester/metadata.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "hasStylesheet": "true"
+}
```
Comment (WuJun2016) [inline assets/brand/harvester/metadata.json:3]: delete this file

File: components/NodeConsumptionGauge.vue [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+};
+</script>
+
+<template>
+  <div class="consumption-gauge">
+    <h3 v-if="resourceName">
+      {{ resourceName }}
+    </h3>
+    <div class="numbers text-muted">
+      <span>{{ t('node.detail.glance.consumptionGauge.used') }}</span> <span>{{ t('node.detail.glance.consumptionGauge.amount', amountTemplateValues) }} <span class="ml-10">/&nbsp;{{ formattedPercentage }}</span></span>
```
Comment (n313893254) [inline components/NodeConsumptionGauge.vue:85]: 换行

File: components/ResourceGauge.vue [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  },
+  computed: {
+    resourceCounts() {
+      return resourceCounts(this.$store, this.resource);
+    },
+    location() {
+      return {
+        name:     'c-cluster-product-resource',
+        params:   { product: VIRTUAL, resource: this.visitResource || this.resource }
```
Comment (n313893254) [inline components/ResourceGauge.vue:65]: components下面的组件，不应该hard-code `VIRTUAL`

File: components/formatter/CloudInitType.vue [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      type:    Object,
+      default: () => {
+        return {};
+      }
+    },
+  },
+
+  computed: {
+    displayLabel() {
+      return this.value?.[HCI.CLOUD_INIT] === 'user' ? 'User Data' : 'Network Data';
```
Comment (n313893254) [inline components/formatter/CloudInitType.vue:16]: i18n

File: components/formatter/HostIp.vue [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    ip() {
+      return this.value;
+    }
+  },
+};
+</script>
+
+<template>
+  <CopyToClipboardText v-if="ip" :text="ip" />
+</template>
```
Comment (n313893254) [inline components/formatter/HostIp.vue:23]: 没有IP显示为空白，是不是用`-`或者文字提示好一些

File: components/form/ConsoleBar.vue [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    return { };
+  },
+
+  computed: {
+    isDown() {
+      return this.isEmpty(this.resource);
+    },
+
+    isRunning() {
+      return this.resource.actualState === 'Running';
```
Comment (DaiYuzeng) [inline components/form/ConsoleBar.vue:29]: Can we use the property `isRunning` of the vm's model to check?

https://github.com/harvester/dashboard/blob/ae7a3caf7fedcb3fb2bf3cabfe583d86781057c1/models/kubevirt.io.virtualmachine.js#L420

File: components/formatter/vmState.vue [hunk 7]
```diff
@@ -0,0 +1,121 @@
+<script>
+import VMState from '@/components/formatter/BadgeStateFormatter';
+import MigrationState from '@/components/formatter/MigrationState';
+// import RestoreProgress from '@/components/formatter/restoreProgress';
```
Comment (n313893254) [inline components/formatter/vmState.vue:4]: Delete unused code

File: components/formatter/vmState.vue [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      }
+    }
+  },
+
+  data() {
+    return { isMigrating: false };
+  },
+
+  computed: {
+    NetworkImpassability() {
```
Comment (n313893254) [inline components/formatter/vmState.vue:39]: 驼峰命名

File: components/formatter/vmState.vue [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        trigger="hover"
+        offset="16"
+      >
+        <span class="tooltip-target">
+          <i class="icon icon-warning icon-lg text-warning" />
+        </span>
+
+        <template slot="popover">
+          <p v-for="(message, index) in warningMessage" :key="message">
+            {{ index+1 }}. {{ message }}
```
Comment (n313893254) [inline components/formatter/vmState.vue:104]: 运算符空格隔开

File: components/formatter/vmState.vue [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    }
+  },
+};
+</script>
+
+<template>
+  <span>
+    <!-- <RestoreProgress :vm="row" /> -->
+    <MigrationState v-show="isMigrating" :vm-resource="row" @state-changed="migrationStateChanged" />
+    <div v-show="!isMigrating" class="state">
```
Comment (n313893254) [inline components/formatter/vmState.vue:91]: 不确定能否使用v-else

File: config/product/virtual.js [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+  });
+
+  // multiVirtualCluster
+  basicType([
+    'cluster-members',
+  ], 'rbac');
+  virtualType({
+    showMenuFun(state, getters, rootState, rootGetters) {
+      return rootGetters['isMultiCluster'];
+    },
```
Comment (n313893254) [inline config/product/virtual.js:88]: showMenu会不会更好，既可以传布尔值，也可以传函数进去。类似于model的属性

File: config/product/virtual.js [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+    weight:     289,
+    route:      {
+      name:     'c-cluster-product-resource',
+      params:   { resource: TEMPLATE }
+    },
+    exact: false,
+  });
+
+  configureType(HCI.BACKUP, {
+    DisableEditInDetail: true,
```
Comment (n313893254) [inline config/product/virtual.js:202]: unused property

File: detail/harvesterhci.io.virtualmachinebackup/basic.vue [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+      return disks.filter((disk) => {
+        return !!disk.cdrom;
+      });
+    },
+
+    flavor() {
+      const domain = this.value?.virtualMachineSpec?.template?.spec?.domain;
+
+      return `${ domain.cpu.cores } vCPU , ${ domain.resources.requests.memory } Memory`;
+    },
```
Comment (n313893254) [inline detail/harvesterhci.io.virtualmachinebackup/basic.vue:77]: 没有使用

File: detail/harvesterhci.io.virtualmachinebackup/index.vue [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    :mode="mode"
+    @apply-hooks="applyHooks"
+  >
+    <Tabbed :side-tabs="true" @changed="onTabChanged">
+      <Tab name="Basics" :label="t('harvester.virtualMachine.detail.tabs.basics')">
+        <OverviewBasics v-if="contentResource" v-model="contentResource.spec.source" mode="view" :memory="memory" />
+      </Tab>
+
+      <Tab
+        name="Volume"
```
Comment (n313893254) [inline detail/harvesterhci.io.virtualmachinebackup/index.vue:100]: 驼峰命名

File: detail/harvesterhci.io.virtualmachineimage.vue [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+  },
+
+  data() {
+    return {};
+  },
+
+  computed: {
+    formattedValue() {
+      return getFileSize(this.value?.status?.size) || '-';
+    },
```
Comment (n313893254) [inline detail/harvesterhci.io.virtualmachineimage.vue:33]: 能不能用formatSi来实现？

```js
formatSi(this.value?.status?.size, {
        increment: 1024,
        maxPrecision: 2,
        suffix:      'B',
        firstSuffix: 'B',
      })
```

File: detail/host/Monitor.vue [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+<template>
+  <div>
+    Monitor
+  </div>
+</template>
+
+<style lang="scss" scoped>
+
+</style>
```
Comment (n313893254) [inline detail/host/Monitor.vue:32]: 这是一个用到的文件吗

File: detail/host/instance.vue [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+</script>
+
+<template>
+  <div class="row host-instances">
+    <div class="col span-12">
+      <SortableTable
+        v-bind="$attrs"
+        :headers="headers"
+        default-sort-by="age"
+        :rows="[...rows]"
```
Comment (n313893254) [inline detail/host/instance.vue:99]: 解构赋值有什么作用吗？ 是不是可以rows=rows

File: components/form/Console.vue [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+      }
+    }
+  },
+
+  computed: {
+    isDown() {
+      return this.isEmpty(this.value);
+    },
+    url() {
+      // const isDev = process.env.dev;
```
Comment (DaiYuzeng) [inline components/form/Console.vue:23]: Remove the comment

File: components/form/SerialConsole/index.vue [hunk 19]
```diff
@@ -0,0 +1,268 @@
+<script>
+import { allHash } from '@/utils/promise';
+import debounce from 'lodash/debounce';
+
+import Socket, {
+  EVENT_CONNECTED,
+  EVENT_CONNECTING,
+  EVENT_DISCONNECTED,
+  EVENT_MESSAGE,
+  //  EVENT_FRAME_TIMEOUT,
```
Comment (DaiYuzeng) [inline components/form/SerialConsole/index.vue:10]: Remove the comment.

File: detail/harvesterhci.io.virtualmachinebackup/keypairs.vue [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+export default {
+  props: {
+    value: {
+      type:     Object,
+      required: true
+    }
+  },
+
+  // async fetch() {
```
Comment (DaiYuzeng) [inline detail/harvesterhci.io.virtualmachinebackup/keypairs.vue:14]: Remove it.

File: detail/kubevirt.io.virtualmachine/annotations-modal/index.vue [hunk 21]
```diff
@@ -0,0 +1,124 @@
+<script>
```
Comment (DaiYuzeng) [inline detail/kubevirt.io.virtualmachine/annotations-modal/index.vue:1]: Remove this component.

File: detail/kubevirt.io.virtualmachine/description-modal/index.vue [hunk 22]
```diff
@@ -0,0 +1,88 @@
+<script>
```
Comment (DaiYuzeng) [inline detail/kubevirt.io.virtualmachine/description-modal/index.vue:1]: Remove this file.

File: detail/kubevirt.io.virtualmachine/labels-modal/index.vue [hunk 23]
```diff
@@ -0,0 +1,101 @@
+<script>
```
Comment (DaiYuzeng) [inline detail/kubevirt.io.virtualmachine/labels-modal/index.vue:1]: Remove this file.

Comment (DaiYuzeng) [reply detail/kubevirt.io.virtualmachine/labels-modal/index.vue:1]: Done.

File: detail/kubevirt.io.virtualmachine/annotations-modal/index.vue (hunk 21 above)
Comment (DaiYuzeng) [reply detail/kubevirt.io.virtualmachine/annotations-modal/index.vue:1]: Done.

File: detail/harvesterhci.io.virtualmachinebackup/keypairs.vue (hunk 20 above)
Comment (DaiYuzeng) [reply detail/harvesterhci.io.virtualmachinebackup/keypairs.vue:14]: Done.

File: components/form/SerialConsole/index.vue (hunk 19 above)
Comment (DaiYuzeng) [reply components/form/SerialConsole/index.vue:10]: Done.

File: components/form/Console.vue (hunk 18 above)
Comment (DaiYuzeng) [reply components/form/Console.vue:23]: Done.

File: components/form/ConsoleBar.vue (hunk 6 above)
Comment (DaiYuzeng) [reply components/form/ConsoleBar.vue:29]: Done.

File: list/harvesterhci.io.dashboard/index.vue [hunk 24]
```diff
@@ -0,0 +1,532 @@
...
+      maxExponent:      99,
+      maxPrecision:     2,
+      minExponent:      0,
+      startingExponent: 0,
+      suffix:           'iB',
+    }
+  }
+};
+
+const METRICS_POLL_RATE_MS = 20000;
```
Comment (DaiYuzeng) [inline list/harvesterhci.io.dashboard/index.vue:39]: Can we use the mixins `metric-poller` to refactor the code?

https://github.com/harvester/dashboard/blob/fed0b43dc9c9041434eb0672fad66e1b581d096a/mixins/metric-poller.js#L6

File: components/NodeConsumptionGauge.vue (hunk 2 above)
Comment (n313893254) [reply components/NodeConsumptionGauge.vue:85]: done

File: components/ResourceGauge.vue (hunk 3 above)
Comment (n313893254) [reply components/ResourceGauge.vue]: Done

File: components/formatter/HostIp.vue (hunk 5 above)
Comment (n313893254) [reply components/formatter/HostIp.vue:23]: LGTM

File: components/formatter/vmState.vue [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+    migrationStateChanged(neu) {
+      this.isMigrating = !!neu;
+    }
+  },
+};
+</script>
+
+<template>
+  <span>
+    <!-- <RestoreProgress :vm="row" /> -->
```
[the remark this answers is not in the document]
Comment (n313893254) [reply components/formatter/vmState.vue:89]: done

File: components/formatter/vmState.vue (hunk 7 above)
Comment (n313893254) [reply components/formatter/vmState.vue:4]: Done

File: components/formatter/vmState.vue (hunk 8 above)
Comment (n313893254) [reply components/formatter/vmState.vue:39]: done

File: components/formatter/vmState.vue (hunk 9 above)
Comment (n313893254) [reply components/formatter/vmState.vue:104]: done

File: config/product/virtual.js (hunk 12 above)
Comment (n313893254) [reply config/product/virtual.js:199]: done

File: detail/harvesterhci.io.virtualmachinebackup/basic.vue (hunk 13 above)
Comment (n313893254) [reply detail/harvesterhci.io.virtualmachinebackup/basic.vue:77]: done

File: detail/harvesterhci.io.virtualmachinebackup/index.vue (hunk 14 above)
Comment (n313893254) [reply detail/harvesterhci.io.virtualmachinebackup/index.vue:100]: done

File: assets/styles/themes/_light.scss [hunk 26]
```diff
@@ -311,7 +311,7 @@ BODY, .theme-light {
   --header-height              : 55px;
   --header-border              : #{$medium};
   --header-border-size         : 1px;
-  --nav-width                  : 230px;
+  --nav-width                  : 260px;
```
[the remark this answers is not in the document]
Comment (n313893254) [reply assets/styles/themes/_light.scss:314]: image

[the remark this answers is not in the document]
Comment (n313893254) [reply assets/styles/themes/_light.scss:314]: 这里需要260px

File: models/node.js [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+
+    const disableMaintenance = {
+      action:     'disableMaintenanceMode',
+      enabled:    this.hasAction('disableMaintenanceMode'),
+      icon:       'icon icon-fw icon-lock',
+      label:      this.t('harvester.action.disableMaintenance'),
+      total:      1
+    };
+
+    if (this.$rootGetters['currentProduct'].inStore === VIRTUAL) {
```
Comment (DaiYuzeng) [inline models/node.js:49]: `_availableActions` should always have a return value

File: detail/kubevirt.io.virtualmachine/annotations-modal/index.vue [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+    name="annotationModal"
+    width="50%"
+    :click-to-close="false"
+    height="auto"
+    class="anno-modal"
+  >
+    <Card>
+      <template #title>
+        <h4 slot="title" class="text-default-text">
+          Edit Annotations
```
Comment (n313893254) [inline detail/kubevirt.io.virtualmachine/annotations-modal/index.vue:84]: i18n

File: pages/c/_cluster/virtual/members/index.vue [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+      :schema="schema"
+      :headers="headers"
+      :rows="filteredClusterRoleTemplateBindings"
+      :groupable="false"
+      :namespaced="false"
+      sub-search="subSearch"
+      :sub-fields="['nameDisplay']"
+    />
+  </div>
+</template>
```
[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply pages/c/_cluster/virtual/members/index.vue:80]: Done.

File: detail/kubevirt.io.virtualmachine/tabs/details/cloud-configs.vue [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+        ref="yamlUser"
+        v-model="cloudData.networkData"
+        class="yaml-editor"
+        :editor-mode="editorMode"
+      />
+    </div>
+  </div>
+</template>
+
+<style lang="scss">
```
Comment (n313893254) [inline detail/kubevirt.io.virtualmachine/tabs/details/cloud-configs.vue:76]: 要不要加scoped?

File: detail/kubevirt.io.virtualmachine/tabs/details/cloud-configs.vue [hunk 31]
```diff
@@ -0,0 +1,98 @@
...
+
+      return out;
+    }
+  },
+
+  methods: {
+    onReady(cm) {
+
+    }
+  },
```
Comment (n313893254) [inline detail/kubevirt.io.virtualmachine/tabs/details/cloud-configs.vue:46]: Remove

File: detail/harvesterhci.io.virtualmachineimage.vue (hunk 15 above)
Comment (DaiYuzeng) [reply detail/harvesterhci.io.virtualmachineimage.vue:33]: Done.

**Pull request closed without merging**